### PR TITLE
include/device.h: Add device class sorting

### PR DESCRIPTION
**Pull request by bjarki-andreasen**
### TL;DR

Adds macro which places struct device API structure in sections based on class, which is then made available to the elf parser, allowing it to determine the class of a struct device without dwarf info or other device meta data. This can be used for sorting through devices by class instead of by node/name using macros

### Disclaimer

This is a working prototype, thats is currently being worked on

### What is the problem?

Right now, we use devicetree aliases to identify hardware using the name only. This ties all struct devices to the node they belong to, and does not ensure that the device driver is of any particular class, or exists at all. From the application, to get the RTC, we currently do this:

```
const struct device *rtc = DEVICE_DT_GET(DT_ALIAS(rtc));
```

This can result in unpredictable behavior, if the device driver is not actually using the RTC API, but the counter API, which we can not protect against outside of userspace currently.

### What will the solution look like?

Knowing all devices that exist by global variable name, and what API they implement (their class), we can do this instead:

`#define DEVICE_CLASS_GET(_class, _index)`

```
const struct device *rtc = DEVICE_CLASS_GET(rtc, 0);
```

Which will return the first instance of any RTC on the system, regardless of where in the devicetree it exists. This macro would ensure that the struct device actually implements the RTC API, and we can even extend the API wrappers to perform an additional validation on the struct device API ptr for an extra layer of protection.

We can additionally check how many instances of some device class there is at compile time, and iterate through them with macros.

`#define DEVICE_CLASS_COUNT(_class)`
`#define DEVICE_CLASS_FOREACH(_class, _fn)`

```
BUILD_ASSERT(DEVICE_CLASS_COUNT(counter) >= 3, "This sample requires at least 3 HW counters");
```

We may become entirely independent from the devicetree nodes the from application (above drivers, kernel and subsystems) using this solution.

### What will change?

A new macro has been added which will be used to define device API implementations within device drivers:

From

```
const struct rtc_driver_api rtc_emul_driver_api = {
        .set_time = rtc_emul_set_time,
        .get_time = rtc_emul_get_time
};
```

To

```
DEVICE_CLASS_API(rtc, rtc_emul_driver_api) = {
        .set_time = rtc_emul_set_time,
        .get_time = rtc_emul_get_time
        ...
};
```

This places the struct device api implementation in the correct section in ROM to be identified and recognized by the elf parser. The following steps are used to determine the class of a device:
1. Build pre linker stage elf file
2. Run elf_parser.py to get list of devices and their properties, including the value of the struct device api pointer
3. Run kernel_device_class.py which uses the API pointer to check which section it points to [rtc_driver_api_area, ...]
4. We now have name and class of all struct devices, DT or none DT
5. FUTURE: create macros to sort through devices by class, like a static version of Linux sysfs

### CODE REVIEW

File: include/zephyr/device.h [hunk 1]
```diff
@@ -737,12 +768,157 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
...
+										\
+			continue;						\
+		}								\
+										\
+		if (api >= TYPE_SECTION_START(_subsys##_driver_api)) {		\
+			break;							\
+		}								\
+										\
+		if (api == NULL) {						\
+			return -ENOSYS;						\
```
Comment (gmarull) [inline include/zephyr/device.h:825]: a macro returning?

**Review by gmarull — CHANGES_REQUESTED** [review]
I'm still not convinced this is the right solution. This again imposes device state sharing, or see it as 1:1 mapping between DT nodes and `struct device`, something I think it's wrong to cover the general case.

There are DT nodes that can map to > 1 device, and they don't need to share the state. This could be the case, for example, of the ST clock/reset controller node. In some MFD, like PMICs, one will likely have subnodes representing actual devices that need their own properties (like a gpio controller, regulator, etc.). These don't need to share the state either, maybe just share a parent to coordinate bus access.

So to summarize, I think we should start thinking about true device classes, that is, expand on top of struct device as Linux does. This would solve, e.g. the 1:1 mapping problem, allow for better common layer handling (now poorly deferred to drivers), etc.

File: include/zephyr/device.h [hunk 2]
```diff
@@ -173,14 +189,29 @@ typedef int16_t device_handle_t;
...
 #define DEVICE_DT_DEFINE(node_id, init_fn, pm, data, config, level, prio, api, \
 			 ...)                                                  \
+	BUILD_ASSERT(IS_EMPTY(__VA_ARGS__), "CONFIG_LIST_SUBSYSTEMS=n");       \
 	Z_DEVICE_STATE_DEFINE(Z_DEVICE_DT_DEV_ID(node_id));                    \
 	Z_DEVICE_DEFINE(node_id, Z_DEVICE_DT_DEV_ID(node_id),                  \
 			DEVICE_DT_NAME(node_id), init_fn, pm, data, config,    \
 			level, prio, api,                                      \
-			&Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_ID(node_id)),     \
-			__VA_ARGS__)
+			&Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_ID(node_id)))
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/device.h:212]: Additional ordinals are not used anywhere in tree except for the unit test, and I really like the simplicity of this small change of the macro :) We can also create a macro like DEVICE_SUBSYS_BUNDLE() which would then be passed to the api ptr if we want to keep the extra ordinals feature

```
DEVICE_SUBSYS(rtc, rtc_api);
DEVICE_SUBSYS(counter, counter_api);
DEVICE_SUBSYS_BUNDLE(api_bundle, &rtc_api, &counter_api);
DEVICE_DT_DEFINE(..., &api_bundle);
```

File: include/zephyr/device.h [hunk 3]
```diff
@@ -737,12 +768,157 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
...
+ * const struct rtc_driver_api *api = DEVICE_SUBSYS_GET(dev, rtc);
+ *
+ * @note Leaves scope returning -ENOSYS if struct device does not
+ * implement specified subsystem.
+ *
+ * @note Subsystem must be defined using DEVICE_SUBSYS() macro
+ *
+ * @return Pointer to implemented struct device subsystem
+ */
+#define DEVICE_SUBSYS_GET(_dev, _subsys)					\
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/device.h:808]: Class could be used to, or API, since its returning a pointer to an api instance :)

File: include/zephyr/device.h [hunk 4]
```diff
@@ -737,12 +768,157 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
...
+	(const struct _subsys##_driver_api *)_dev->api;				\
+	STRUCT_SECTION_START_EXTERN(_subsys##_driver_api);			\
+	STRUCT_SECTION_END_EXTERN(_subsys##_driver_api);			\
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/device.h:811]: Yeah, its not exactly pretty :) It will probably be replaced with an inline function like
`inline const struct rtc_driver_api *rtc_api_get(const struct device * dev)` to be optimized by the compiler.

The ugliness is there to hopefully make it as fast as possible for testing :)

File: include/zephyr/device.h (hunk 1 above)
Comment (bjarki-trackunit) [reply include/zephyr/device.h:825]: @gmarull State sharing (from subsystem listing) is intended for devices like I3C, which inherit existing APIs. I had to use existing APIs for the testing :)  It could also be used for devices like UARTs, which have three quite different APIs in one, these could be single APIs instead (poll, isr, async), which is a bit cleaner IMO and potentially  more memory efficient.

I agree we should have true device classes, giving us more generic parent/subdevice management to handle PM and access synchronization, but that's a topic of its own I think :)

**Comment by bjarki-trackunit** [comment]
@gmarull State sharing (from subsystem listing) is intended for devices like I3C, which inherit existing APIs. I had to use existing APIs for the testing :) It could also be used for devices like UARTs, which have three quite different APIs in one, these could be single APIs instead (poll, isr, async), which is a bit cleaner IMO and potentially more memory efficient.

I agree we should have true device classes, giving us more generic parent/subdevice management to handle PM and access synchronization, but that's a topic of its own I think :)

**Comment by bjarki-trackunit** [comment]
I have updated the draft to use the elf_parser to determine the class of a device (which driver API it implements) and generate a header file from this information. Example output

```
#define DEVICE_P_CLASS___device_dummy_rtc0 rtc
#define DEVICE_P_CLASS___device_dummy_counter0 counter
#define DEVICE_P_CLASS___device_dummy_bbram0 bbram
```
